### PR TITLE
skills: add basic grant-revoke methods

### PR DESCRIPTION
**Pull request by zyga**
This branch adds the four basic grant / revoke methods

Repository.Grant() grants a skill to a slot.
Repository.Revoke() revokes an existing grant.
Repository.GrantedTo() returns information about all skills granted to particular slots in a given snap.
Repository.GrantedBy() returns information about all skills, from a particular snap, granted to various slots in various snaps.

There are several related internal changes. Removing used skills or
occupied slots is no longer allowed and now returns an error.
    
Internally tests were tweaked to ensure that test skill and test slot
don't belong to the same snap. This helps in making asymmetric tests for
granted-to vs granted-by easier to follow.

### CODE REVIEW

File: skills/repo.go [hunk 1]
```diff
@@ -33,14 +33,20 @@ type Repository struct {
 	m     sync.Mutex
 	types map[string]Type
 	// Indexed by [snapName][skillName]
-	skills map[string]map[string]*Skill
+	skills          map[string]map[string]*Skill
+	slots           map[string]map[string]*Slot
+	skillUsedBySlot map[*Slot]map[*Skill]bool
+	slotsUsingSkill map[*Skill]map[*Slot]bool
```
Comment (niemeyer) [inline skills/repo.go]: `skillUsedBySlot` => `slotSkills`
`slotsUsingSkill` => `skillSlots`

File: skills/repo.go [hunk 2]
```diff
@@ -157,6 +168,191 @@ func (r *Repository) RemoveSkill(snapName, skillName string) error {
...
+		return fmt.Errorf("cannot grant skill %q from snap %q, no such skill", skillName, skillSnapName)
+	}
+	// Ensure that such slot exists
+	slot := r.slots[slotSnapName][slotName]
+	if slot == nil {
+		return fmt.Errorf("cannot grant skill to slot %q from snap %q, no such slot", slotName, slotSnapName)
+	}
+	// Ensure that skill and slot are compatible
+	if slot.Type != skill.Type {
+		return fmt.Errorf("cannot grant skill %q from snap %q to slot %q from snap %q, skill type %q doesn't match slot type %q",
```
Comment (niemeyer) [inline skills/repo.go]: In these cases I think we should go short-hand, with "<snap>:<skill or slot>":

`cannot grant skill "%s:%s" (skill type %q) to "%s:%s" (skill type %q)`

The names were validated, so we can trust them to not contain funky chars.

File: skills/repo.go [hunk 3]
```diff
@@ -157,6 +168,191 @@ func (r *Repository) RemoveSkill(snapName, skillName string) error {
...
+			skillName, skillSnapName, slotName, slotSnapName, skill.Type, slot.Type)
+	}
+	// Ensure that slot and skill are not connected yet
+	if r.skillUsedBySlot[slot][skill] {
+		return fmt.Errorf("cannot grant skill %q from snap %q to slot %q from snap %q twice",
```
Comment (niemeyer) [inline skills/repo.go]: This should probably not be an error. The exact semantics the function call requested are already in place, so there's no reason to blow it up I think. We should just return nil, and the call sites will be happy that the state they want is active.

File: skills/repo.go [hunk 4]
```diff
@@ -157,6 +168,191 @@ func (r *Repository) RemoveSkill(snapName, skillName string) error {
...
+	if slot == nil {
+		return fmt.Errorf("cannot revoke skill from slot %q from snap %q, no such slot", slotName, slotSnapName)
+	}
+	// Ensure that slot and skill are connected
+	if !r.skillUsedBySlot[slot][skill] {
+		return fmt.Errorf("cannot revoke skill %q from snap %q from slot %q from snap %q, it is not granted",
+			skillName, skillSnapName, slotName, slotSnapName)
+	}
+	delete(r.skillUsedBySlot[slot], skill)
+	delete(r.slotsUsingSkill[skill], slot)
```
Comment (niemeyer) [inline skills/repo.go]: As in skill and slot removals, the base maps themselves should have the keys deleted once the respective value is empty, so they don't grow out of bounds.

**Comment by niemeyer** [comment]
LGTM once you're happy about these points.

**Comment by zyga** [comment]
Refreshed, please have a 2nd look.

**Comment by chipaca** [comment]
:+1:

**Pull request merged**